pkg/apis/storage/types.go

```diff
@@ -293,8 +293,34 @@ type CSIDriverSpec struct {
 	// "csi.storage.k8s.io/pod.uid": string(pod.UID)
 	// +optional
 	PodInfoOnMount *bool
+
+	// If set, Mode defines how this CSI volume driver can be used. The default
+	// is "persistent", which is the usage defined by the CSI specification.
+	// Other modes are:
+	// - "ephemeral": only use this driver for inline ephemeral volumes (CSIVolumeSource)
+	// - "persistent+ephemeral": both modes supported
+	// More modes may be added in the future.
+	// +optional
+	Mode *CSIDriverMode
 }
 
+// CSIDriverMode is an enumeration of possible usage modes of a CSI driver.
+// More modes may be added in the future.
+type CSIDriverMode string
+
+const (
+	// PersistentDriverMode explicitly confirms that the driver implements
+	// the full CSI spec. It is the default when CSIDriverSpec.Mode is not
+	// set.
```

Comment:
Add some additional details here about how the naming of the mode relates to other "persistent" concepts in Kube - i.e. from discussion in api review "persistent" CSI driver modes are intended to relate to a "persistent volume claims" while "ephemeral" inline volumes do not.  I would also suggest adding a few examples of what sorts of drivers would use this mode (so that an author / casual browser can orient themselves).

so for example for ephemeral "a CSI driver that created/downloaded a vault secret and put it into a directory for a given pod might use ephemeral driver mode because they want that secret to be tied to the pod, and ephemeral gives them the ability to ...".  And then you might want to also mention that the driver is free to have a broader scope than the life of the pod, but that might be less common - "for example, an ephemeral inline CSI volume might share a temporary, garbage-collected empty dir that is only accessible to pods in one namespace".

Reply:
Are folks going to be reading source code for this? I'm thinking this might be better done as an extension to the KEP?

Reply:
Let's document this properly in the API and then update the KEP accordingly. Then whatever people encounter first has a good description.

Reply:
The comments in the external types are used to generate api docs.  And I believe 'kubectl explain' uses this too.  So it's good to have these comments be very thorough.

Reply:
Ah, ok. That makes sense.